ckan/logic/action/create.py

```diff
@@ -619,6 +619,77 @@ def member_create(context, data_dict=None):
     return model_dictize.member_dictize(member, context)
 
 
+def package_member_create(context, data_dict):
+    '''Make a user a collaborator in a dataset.
+
+    If the user is already a collaborator in the dataset then their
+    capacity will be updated.
+
+    Currently you must be an Admin on the dataset owner organization to
+    manage collaborators.
+
+    Note: This action requires the collaborators feature to be enabled with
+    the :ref:`ckan.auth.allow_dataset_collaborators` configuration option.
+
+    :param id: the id or name of the dataset
+    :type id: string
+    :param user_id: the id or name of the user to add or edit
+    :type user_id: string
+    :param capacity: the capacity or role of the membership. Must be one of
+        "editor" or "member". Additionally
+        if :ref:`ckan.auth.allow_admin_collaborators` is set to True, "admin"
+        is also allowed.
+    :type capacity: string
+
+    :returns: the newly created (or updated) collaborator
+    :rtype: dictionary
+    '''
+
+    model = context['model']
+
+    package_id, user_id, capacity = _get_or_bust(
+        data_dict,
+        ['id', 'user_id', 'capacity']
+    )
+
+    allowed_capacities = authz.get_collaborator_capacities()
+    if capacity not in allowed_capacities:
+        raise ValidationError(
+            _('Role must be one of "{}"').format(', '.join(
+                allowed_capacities)))
+
+    package = model.Package.get(package_id)
+    if not package:
+        raise NotFound(_('Dataset not found'))
+
+    user = model.User.get(user_id)
+    if not user:
+        raise NotFound(_('User not found'))
```

Comment:
During last year we've merged few PRs which were trying to prevent exposing of the user account to anonymous request. For example, you won't see **User not found** when resetting the password, you'll see **email was sent to your mailbox**  even if you misspelled username. But this error comes before access check and creates a way to verify, whether the account exists on the portal.

Reply:
Good point. What error would you suggest returning here? Something generic like "Object not found" for both missing datasets and users is not very useful as it's trivial to check if a (public) dataset exists. So maybe raise `NotAuthorized` if the user is not found?

Reply:
I think it's optimal solution, so +1 for `NotAuthorized`

Reply:
`ValidationError` seems more apt for a problem with one parameter of an API call, but I agree it should be done after the `_check_access` call below

Reply:
@wardi the problem I see with `ValidationError` is that is generally tied to one or more fields values (eg "id is missing", "value not allowed", etc) and I'm not sure what error message we would include with one of them (`ValidationError: User not authorized` seems wrong). I went for a NotAuthorized but happy to change it if you have an alternative.